src/analysis/convex/cone/proper.lean

```diff
@@ -56,82 +57,79 @@ lemma closure_eq_iff_is_closed {K : convex_cone ℝ E} : K.closure = K ↔ is_cl
 
 end convex_cone
 
-section definitions
-
 /-- A proper cone is a convex cone `K` that is nonempty and closed. Proper cones have the nice
 property that the dual of the dual of a proper cone is itself. This makes them useful for defining
 cone programs and proving duality theorems. -/
-structure proper_cone (E : Type*) [normed_add_comm_group E] [inner_product_space ℝ E]
-  extends convex_cone ℝ E :=
+structure proper_cone (𝕜 : Type*) (E : Type*)
+  [is_R_or_C 𝕜] [ordered_semiring 𝕜] [normed_add_comm_group E] [inner_product_space 𝕜 E]
```

Comment:
Isn't `[is_R_or_C 𝕜] [ordered_semiring 𝕜]` a very bad idea?  I think that means "put two unrelated `*` operations on `𝕜`".

Reply:
This is the only way I can make `proper_cone k E` work. Should I just get rid of this and leave it as `proper_cone E`? Almost all the results rely on the [dual cone section](https://leanprover-community.github.io/mathlib_docs/analysis/convex/cone/basic.html#the-dual-cone) anyways which is only for real cones. So this generality isn't very useful.

Reply:
I think the correct approach is to remove `[ordered_semiring k]` here and use the reals whenever you need them in the lemmas.

Reply:
In mathlib we really like to have the definitions and theorems as general as possible, even if they are a bit useless in the most general form.

Reply:
I think I agree here that adding this `𝕜` argument is pretty useless.

You could justify it if you changed `inner_dual_cone` to

```lean
variables (R : Type*) {H : Type*} [normed_add_comm_group H] [inner_product_space ℝ H]
  [ordered_semiring R] [has_smul R H] [smul_with_zero R ℝ] [ordered_smul R ℝ]
  [is_scalar_tower R ℝ ℝ]
  [is_scalar_tower R ℝ H] (s t : set H)
open_locale real_inner_product_space

/-- The dual cone is the cone consisting of all points `y` such that for
all points `x` in a given set `0 ≤ ⟪ x, y ⟫`. -/
def set.inner_dual_cone (s : set H) : convex_cone R H :=
{ carrier := { y | ∀ x ∈ s, 0 ≤ ⟪ x, y ⟫ },
  smul_mem' := λ c hc y hy x hx,
  begin
    rw [←smul_one_smul ℝ c y, real_inner_smul_right, smul_mul_assoc, one_mul],
    exact smul_nonneg hc.le (hy x hx)
  end,
  add_mem' := λ u hu v hv x hx,
  begin
    rw inner_add_right,
    exact add_nonneg (hu x hx) (hv x hx)
  end }
```
but this doesn't seem very useful either

Reply:
The issue is that [convex_cone](https://leanprover-community.github.io/mathlib_docs/analysis/convex/cone/basic.html#convex_cone) requires an `ordered_semiring` structure whereas [inner product_space](https://leanprover-community.github.io/mathlib_docs/analysis/inner_product_space/basic.html#inner_product_space) requires `is_R_or_C`

Reply:
My suspicion is that this is one of those rare theories that only make sense over `R` because we need both an order structure and some analytic structure. 

I'll put the `R` back and leave a note about this discussion. Changing for `R` to `k` was a simple find and replace so it would be very easy to revert back if a better solution comes along.

Reply:
Removing inner_product_space fixed this :)
I'll weaken the hypothesis as much as possible and push an update.